File: python-leetcode/linked-list/142.环形链表 II.py
from typing import Optional

from type import ListNode


class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return None

        slow = fast = head

        while slow and fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                # 相遇，找环入口
                x = head
                y = slow
                while x != y:
                    assert x is not None and y is not None  # 保证不为 None
                    x = x.next
                    y = y.next
                return x
        return None

File: go-leetcode/linked-list/142.环形链表 II.go
package leetcode

func detectCycle(head *ListNode) *ListNode {
	if head == nil {
		return nil
	}

	slow, fast := head, head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next

		if slow == fast {
			// 相遇，找环入口
			x := head
			y := slow

			for x != y {
				x = x.Next
				y = y.Next
			}

			return x
		}
	}

	return nil
}
